test/cluster/basic_test.go

```diff
@@ -101,3 +101,348 @@ func TestBasicCluster(t *testing.T) {
+			require.Contains(t, err.Error(), fmt.Sprintf("referrer: '%s/data/tx'", srv.URL()))
+			txRedirected++
+			followerServer = srv
+			followerValue = i
+		} else {
+			t.Logf("tx submitted: %s, %+v", txID, rcpt)
+			txOK++
+			leaderServer = srv
+			t.Logf("Leader is: %s", srv.URL())
+		}
+	}
+	require.Equal(t, 1, txOK)
+	require.Equal(t, 2, txRedirected)
+
+	require.NoError(t, c.ShutdownServer(leaderServer))
+
+	txID, rcpt, err := followerServer.WriteDataTx(t, worldstate.DefaultDBName, "bob", []byte{uint8(followerValue), uint8(followerValue)})
+	if err != nil {
+		t.Logf("txID: %s, error: %s", txID, err)
+	} else {
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+	}
+
+	require.NoError(t, c.StartServer(leaderServer))
+
+	time.Sleep(10 * time.Second)
+
+	dataEnv, err := leaderServer.QueryData(t, worldstate.DefaultDBName, "bob")
+	require.NoError(t, err)
+	dataResp := dataEnv.GetResponse()
+	t.Logf("data: %+v", dataResp)
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - shutting down and starting all servers in the cluster.
+func TestClusterRecovery(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	require.NoError(t, c.Restart())
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - submit a tx to each of the 3, with a client that catches redirects.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - repeat until each server was a leader at least once.
+func TestAllNodesGetLeadership(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	time.Sleep(10 * time.Second) //TODO replace by a call to detect a leader
+
+	var txOK int
+	leaders := [3]int{0, 0, 0}
```

Comment:
you use `int` as a `bool` right? why not just `leaders := [3]bool{false, false, false}` ? seems more readable

Reply:
Done